### PR TITLE
Add function igraph_is_same_graph for tests

### PR DESCRIPTION
**Pull request by iosonofabio**
Addressing #1599.

This relies on `graph->ii` which is a vector of edge ids, double sorted by source and target vertices. If that representation goes, we should adapt this function. The easiest way to do that would be to sort the edges ourselves in the same way inside the function.

edit: I should put a header somewhere. Which header file would be appropriate for this to be considered internal?

### CODE REVIEW

**Comment by vtraag** [comment]
> edit: I should put a header somewhere. Which header file would be appropriate for this to be considered internal?

I believe all other functions that work directly on the `igraph_t` object are stored in `igraph_interface.h`.

**Comment by iosonofabio** [comment]
two quick questions:

1. do we want this function to return `igraph_bool_t` instead of `int`?
2. I haven't learned how to run tests locally yet with the new CMake system. I tried to follow the pattern in `.github/workflows/cmake...` but it says 0% tests passed. Sorry about that, having a working local test suite would reduce the number of commits in these PRs.

**Comment by iosonofabio** [comment]
hmm the test failure seems to be due to *memory leaks in the erdos-renyi test*? I'm not sure what's going on here... @ntamas?

**Comment by vtraag** [comment]
> do we want this function to return `igraph_bool_t` instead of `int`?

That's probably a good idea.

> I haven't learned how to run tests locally yet with the new CMake system.

You did compile the tests first with `cmake --build . --target check` (or simply `make check`)? Running them should be done using `ctest`, possibly including a configuration, e.g. `ctest -C Debug`, but I think that is Windows only.

**Comment by iosonofabio** [comment]
Thanks, now it returns `igraph_bool_t`. I got it to test locally by using:

```
mkdir build
cd build
cmake ..
make
make check
```
and that already tests it. All good here.

**Comment by vtraag** [comment]
> that already tests it.

Ah, so `make check` both builds and runs the tests, even better. Just FYI, with `ctest` you can run specific tests with a regex specifier, e.g. `ctest -R is_same_graph`. That way, you can easily test only one test case. SImilarly, you can build only that test by invoking `cmake --build . --target test_igraph_is_same_graph`.

File: src/graph/type_indexededgelist.c [hunk 1]
```diff
@@ -1691,3 +1691,62 @@ int igraph_incident(const igraph_t *graph, igraph_vector_t *eids,
...
+
+/**
+ * \function igraph_is_same_graph
+ * \brief Check if two graphs are topologically the same
+ *
+ * \param graph1 The first graph object.
+ * \param graph2 The second graph object.
+ * \return 1 if they are the same, 0 if they are different.
+ *
+ * NOTE: this function is mostly for internal use in unit tests.
```
Comment (szhorvat) [inline src/graph/type_indexededgelist.c:1704]: Is there any good reason for this? If the function is robust, its use should not be discouraged (and it _should_ be robust). Testing is the most common, but by far not the only use case.

There are several other changes I want to make to the docs. Is it okay if I make edits to this text on this branch?

**Comment by iosonofabio** [comment]
There is a reason, which is that it is the truth ;-)

But sure edit at will, nothing personal

**Comment by ntamas** [comment]
> do we want this function to return igraph_bool_t instead of int?

The rule of thumb is: if the function can fail, ever, then it should return an `int` with an igraph error code (zero if everything was OK), and the _real_ result should be returned in an output argument. Failure can mean things like "not enough memory to allocate auxiliary data structures" as well. So, if we decide to return `igraph_bool_t` here, we commit ourselves to never use memory allocations inside the function, and to never use any igraph function that can fail on its own. I'm okay with that, just wanted to make sure that you are aware of it.

**Comment by ntamas** [comment]
I'm not particularly happy with the direct usage of `graph->ii`, but there might be valid reasons for that.

Roughly speaking, anything that accesses the internals of an `igraph_t` object is considered to be part of the "basic graph interface". We are aiming for this interface to be as small as possible in order to make it easier for us to experiment with alternative graph representations. My plans for the future are to make the basic graph interface smaller, not larger; we already have several methods in the basic graph interface that could be moved outside:

* `igraph_empty()` simply calls `igraph_empty_attrs()`, and I cannot really conceive any graph data structure where this would not be true, so `igraph_empty()` could be moved outside the basic API, leaving only `igraph_empty_attrs()` in there.

* `igraph_delete_vertices()` delegates everything to `igraph_delete_vertices_idx()` so this could be removed from the basic API as well.

...and so on, and so on.

The point I'm trying to make here is that if we decide to reach into the internals of `igraph_t` in `igraph_is_same_graph()`, and then later on we decide to implement a new graph data structure to see whether we gain something in terms of efficiency, we would need to deal with the implementation of `igraph_is_same_graph()` _again_ with this new graph data structure.

**Comment by ntamas** [comment]
By the way, I have seen the comment from @vtraag in the original issue where he advocates for this function being part of the core graph API. I'm okay with that as well, just wanted to make sure that we understand the implications. Also, if we do decide that this function becomes part of the core graph API, then we should make sure to document very carefully what the function needs to do (compare directednes, compare number of nodes, ensure that the edge sets are the same, not necessarily in the same order, but with same edge multiplicities -- so basically a "multiset equality"), and what the function does not need to do (e.g., compare attributes). This is to help future implementors of other graph data structures to know what to deal with and what not to deal with.

**Comment by iosonofabio** [comment]
Yes @ntamas that is clear. I agree with the goal to reduce the basic interface.

Double sorted edges are useful for this function, whether (1) they are sorted using an internal data structure like current `ii` or (2) by calling a public function to get the sorted edge ids (which is how `ii` is made in the first place, o'course).

If you prefer option 2 then such a public function needs to be written. Its implementation would currently rely on `ii` for speed but would need to change if we drop `ii`. Or we could make it slower and recompute the sorted edges pretending that we don't know about `ii`.

The bottom line is that if `ii` is just a source of worry we should just dump it and consider it deprecated. But if it's useful for internal functions like this then there's nothing wrong in using it.

Btw @szhorvat this whole discussion is exactly why I added the documentation line warning users that this function is internal.

**Comment by iosonofabio** [comment]
I'm thinking that if this function creates such headaches in the public API, we could move it to the file that contains other test utility functions and be done with it

**Comment by vtraag** [comment]
I think we need a more precise terminology. Yes, this function `igraph_is_same_graph` should be considered part of the core graph API, as @ntamas formulated it. This is what I referred to as "internal" previously, but this is a misleading nomenclature it now seems, which is causing some confusion. It should be a *public* function, that can be used by users of the library, hence the declaration should be made in `include/igraph_interface.h`, as it is done now. If the function is *not* public, it should *not* be declared in `include/igraph_interface.h`. Hence, the comment by @szhorvat that we should not state in the documentation that this function is only for internal use, which I agree with.

It does imply, as @ntamas already observed, that this implementation is part of the core graph API, and hence, should be changed along with a different implementation.

Although it could technically be possible to move it outside the current graph representation, I think that would sacrifice too much on speed. At the moment the graph is already internally sorted, and you essentially need to perform exactly this operation in order to identify whether two graphs are identical (i.e. the exact same edges, even if the original edge ordering is different). That was the original reason why I proposed it should be part of the core graph API.

**Comment by vtraag** [comment]
> if we decide to return `igraph_bool_t` here, we commit ourselves to never use memory allocations inside the function, and to never use any igraph function that can fail on its own. I'm okay with that, just wanted to make sure that you are aware of it.

Yes, this is a good point. I assumed that it could never fail because it does not call any other functions that can fail at the moment. However, that is not necessarily the case for future representations, because they are able to fail, for example because they do need to sort some edges. For that reason, I think you are correct, and the API should be declared with returning an `int`, and the result should be provided in a pointer/reference.

**Comment by szhorvat** [comment]
A lot has happened here while I was out. In short, here are my preferences:

 - During the winter break I did think a bit about whether to make this part of the core API (i.e. messing with internal data structures), but didn't reach a definitive conclusion. **I am weakly in favour of making it part of the core API,** for performance reasons.
 - **I am strongly in favour of returning an error code** instead of returning a bool. Especially for the core API, stability is paramount, and it is not at all unlikely that if we change the graph data structure, it will become possible for this function to fail.

**Comment by szhorvat** [comment]
> hmm the test failure seems to be due to _memory leaks in the erdos-renyi test_? I'm not sure what's going on here... @ntamas?

This is now fixed (it wasn't a leak, but there was a problem, see #1606).

**Comment by iosonofabio** [comment]
Aaargh, I was pushing my changes at the same time...

**Comment by szhorvat** [comment]
> Aaargh, I was pushing my changes at the same time...

I'm sorry about that, I wanted to push these before you woke up, but my internet stopped working for a while ... 

Feel free to overwrite my version (but please keep the docs).

**Comment by iosonofabio** [comment]
Alright, it's too messy to have two people editing the code in parallel.

@szhorvat please ask before changing the code in my PRs next time. My understanding was that you were going to change the documentation text, not a bunch of other things too. In any case, please keep pushing here and then we can merge, I'm stopping my contribution to this PR now.

And error codes are ints, not `igraph_bool_t`

**Comment by iosonofabio** [comment]
@szhorvat in the tests, you still have to assert that the error code returned by the function is zero, to avoid issues in the future.

**Comment by szhorvat** [comment]
I'm sorry, my intention was to merge it. I considered this to be basically complete save for documentation and a tedious detail about how it returns it result.

**Comment by iosonofabio** [comment]
that's ok, but please add those asserts in the tests if that's our practice now.

**Comment by szhorvat** [comment]
> that's ok, but please add those asserts in the tests if that's our practice now.

That's done now. Can you take a look with a fresh morning mind and if you're happy and the CI tests pass, then squash-and-merge it?

**Comment by iosonofabio** [comment]
Thank you @szhorvat!

It looks good to me.

@ntamas is the current practice to squash and merge or just merge? I remember a previous conversation on this topic but can't quite recall what the conclusion was, sorry about that

**Comment by szhorvat** [comment]
> is the current practice to squash and merge or just merge?

It depends on the situation. When the commits are messy, and the commit messages assume a specific context that won't be apparent when reading them a year later, it should be squash-and-merge. I find it easier to work this way as I can just push many commits without having to keep all of them organized. The tradeoff is that the PR should be relatively small.

Sometimes, when the PR is too big, it's better to keep commits separate. But then the PR author should make sure that the commits are all nice and well-structured, which can be a pain. This is the case in PR #1583 (and I very much regret choosing this path because it's a pain)

**Comment by iosonofabio** [comment]
Personally I never squash because you can see in the git DAG that those commits all came together by the same author(s) anyway, and a git DAG is not a showroom, I'd rather have it a little verbose than too terse.

Anyway, I don't usually merge my own PRs out of courtesy, @ntamas feel free to do either one, your call.

**Review by ntamas — APPROVED** [review]

**Comment by ntamas** [comment]
Thanks a lot guys!
